sparse/impl/KokkosSparse_gmres_numeric_impl.hpp

```diff
@@ -0,0 +1,388 @@
+//
+// Questions? Contact Siva Rajamanickam (srajama@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOSSPARSE_IMPL_GMRES_NUMERIC_HPP_
+#define KOKKOSSPARSE_IMPL_GMRES_NUMERIC_HPP_
+
+/// \file KokkosSparse_gmres_numeric_impl.hpp
+/// \brief Implementation(s) of the numeric phase of GMRES.
+
+#include <KokkosKernels_config.h>
+#include <Kokkos_ArithTraits.hpp>
+#include <KokkosSparse_gmres_handle.hpp>
+#include <KokkosBlas.hpp>
+#include <KokkosBlas3_trsm.hpp>
+#include <KokkosSparse_spmv.hpp>
+//#include <KokkosSparse_Preconditioner.hpp>
+#include "KokkosKernels_Error.hpp"
+
+//#define NUMERIC_OUTPUT_INFO
+
+namespace KokkosSparse {
+namespace Impl {
+namespace Experimental {
+
+template <class GmresHandle>
+struct GmresWrap {
+  //
+  // Useful types
+  //
+  using execution_space         = typename GmresHandle::execution_space;
+  using index_t                 = typename GmresHandle::nnz_lno_t;
+  using size_type               = typename GmresHandle::size_type;
+  using scalar_t                = typename GmresHandle::nnz_scalar_t;
+  using HandleDeviceEntriesType = typename GmresHandle::nnz_lno_view_t;
+  using HandleDeviceRowMapType  = typename GmresHandle::nnz_row_view_t;
+  using HandleDeviceValueType   = typename GmresHandle::nnz_value_view_t;
+  using HandleDevice2dValueType = typename GmresHandle::nnz_value_view2d_t;
+  using karith                  = typename Kokkos::ArithTraits<scalar_t>;
+  using device_t                = typename HandleDeviceEntriesType::device_type;
+
+  /**
+   * The main gmres numeric function. Copied with slight modifications from
+   * example/gmres/gmres.hpp
+   */
+  template <class AMatrix, class BType, class XType>
+  static void gmres_numeric(GmresHandle& thandle, const AMatrix& A,
+                            const BType& B, XType& X) {
+    using ST = typename karith::val_type;  // So this code will run with
+                                           // scalar_t = std::complex<T>.
+    using MT                  = typename karith::mag_type;
+    using HandleHostValueType = typename HandleDeviceValueType::HostMirror;
+
+    ST one  = karith::one();
+    ST zero = karith::zero();
+
+    Kokkos::Profiling::pushRegion("GMRES::TotalTime:");
+
+    // Store solver options:
+    const auto n          = thandle.get_nrows();
+    const int m           = thandle.get_m();
+    const auto maxRestart = thandle.get_max_restart();
+    const auto tol        = thandle.get_tol();
+    const auto ortho      = thandle.get_ortho();
+    const auto verbose    = thandle.get_verbose();
+    auto precond          = thandle.get_precond();
+
+    bool converged     = false;
+    size_type cycle    = 0;  // How many times have we restarted?
+    size_type numIters = 0;  // Number of iterations within the cycle before
+    // convergence.
+    MT nrmB, trueRes, relRes, shortRelRes;
+
+    if (verbose) {
```

Comment:
I reorganized the output and the local view definitions a bit, otherwise it's the same as gmres.hpp.